### PR TITLE
Pipeline for Kubernetes CI

### PR DESCRIPTION
**Pull request by accassid**
* Adjusting docker image to work with kubernetes pods
* Adjusting production settings to work with kubernetes pods
* Writing CI steps to build, test, deploy to the ECR, migrate and turn over pods in separate steps
* Fixing failing tests

### CODE REVIEW

File: backend/.dockerignore [hunk 1]
```diff
@@ -2,10 +2,10 @@
 **/*.sqlite3
 docker-*.yml
 Dockerfile
+Dockerfile.*
 **/*.md
 nginx/
 __pycache__
 ./temp
 *_data.json
-**/test*.py
```
Comment (ldeluigi) [inline backend/.dockerignore:10]: Why aren't you ignoring tests? They are not necessary for the application to work, while the weigh in the final image.

File: .github/workflows/ci.yml
Comment (ldeluigi) [inline .github/workflows/ci.yml:1]: This file should be refactored once we establish the right workflow, restoring semantic versioning automation

File: .gitignore [hunk 2]
```diff
@@ -142,3 +142,5 @@ temp/
 # Localhost certificates
 *.crt
 *.key
+
```
Comment (ldeluigi) [inline .gitignore:145]: You should add a little comment/title like the rest of the gitignore

File: backend/.kubernetes/kubernetes.yaml [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      app: spellbook-api
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: spellbook-api
+    spec:
+      containers:
+        - name: spellbook-api-app
+          image: 083767677168.dkr.ecr.us-east-2.amazonaws.com/spellbook-prod-ecr
```
Comment (ldeluigi) [inline backend/.kubernetes/kubernetes.yaml:26]: Shouldn't we hide this link?
Alternatively, we could directly publish an official Docker Hub image as this project is open source, what do you think?

File: backend/.kubernetes/kubernetes.yaml [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  type: NodePort
+  selector:
+    app: spellbook-api
+  ports:
+    - port: 80
+      targetPort: 80
+
+---
+# Ingress Load Balancer
+apiVersion: networking.k8s.io/v1
```
Comment (ldeluigi) [inline backend/.kubernetes/kubernetes.yaml:117]: If this load balancer replaces nginx, whos i in charge of serving static files?

File: backend/.kubernetes/migration/kustomization.yaml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+resources:
+- kube.migration.yaml
```
Comment (ldeluigi) [inline backend/.kubernetes/migration/kustomization.yaml:2]: This file should end with a new line

File: backend/Dockerfile [hunk 6]
```diff
@@ -52,17 +50,20 @@ RUN pip install --no-cache psycopg[binary] gunicorn
 # copy project
 COPY --chown=app:app . $APP_HOME
 
-# change to the app user
-USER app
+# # change to the app user
+# USER app
```
Comment (ldeluigi) [inline backend/Dockerfile:54]: Why do you need to run the application as root?

File: backend/Dockerfile [hunk 7]
```diff
@@ -52,17 +50,20 @@ RUN pip install --no-cache psycopg[binary] gunicorn
...
 
 ARG SECRET_KEY
 ENV SECRET_KEY=${SECRET_KEY}
 ENV DJANGO_SETTINGS_MODULE=backend.production_settings
 
+# RUN python manage.py collectstatic --no-input --clear
```
Comment (ldeluigi) [inline backend/Dockerfile:60]: How are static files handled now?
Following this guides:
* https://docs.djangoproject.com/en/4.2/howto/static-files/#configuring-static-files
* https://docs.djangoproject.com/en/4.2/howto/static-files/deployment/#serving-static-files-in-production

It's probably a good idea to serve generated, static files from a different source than the web server.
I don't know how to handle this task though.

File: backend/Dockerfile [hunk 8]
```diff
@@ -52,17 +50,20 @@ RUN pip install --no-cache psycopg[binary] gunicorn
...
+# RUN sleep 3
+
 # run entrypoint.prod.sh
-EXPOSE 8000
+EXPOSE 80
```
Comment (ldeluigi) [inline backend/Dockerfile:64]: The port should match the same used in line 66 of this file, which is currently 8000.

File: backend/Dockerfile [hunk 9]
```diff
@@ -52,17 +50,20 @@ RUN pip install --no-cache psycopg[binary] gunicorn
...
 ENTRYPOINT ["/bin/sh", "entrypoint.prod.sh"]
-RUN echo -e 'gunicorn backend.wsgi:application --bind 0.0.0.0:8000\n' > entrypoint.prod.sh
+RUN echo -e 'gunicorn backend.wsgi:application --bind 0.0.0.0:8000 --workers=3\n' > entrypoint.prod.sh
```
Comment (ldeluigi) [inline backend/Dockerfile:66]: Isn't 3 workers arbitrary?
Maybe you should leave it as default, or else make it configurable with some variable?

File: backend/Dockerfile [hunk 10]
```diff
@@ -71,8 +72,11 @@ RUN echo -e 'gunicorn backend.wsgi:application --bind 0.0.0.0:8000\n' > entrypoi
 
 FROM base AS demo
 
+ADD nginx.conf /etc/nginx/nginx.conf
```
Comment (ldeluigi) [inline backend/Dockerfile:75]: Is nginx really needed here? I think that Nginx should be spawned as a separate docker container, just like the docker-compose.yml was doing.

File: backend/Dockerfile.kubernetes
Comment (ldeluigi) [inline backend/Dockerfile.kubernetes:1]: Is this file really needed? I feel it's a repetition of the Dockerfile second stage. We could probably find and group the common layers into having a single, multipurpose and multistaged Dockerfile called simply Dockerfile.

File: backend/backend/production_settings.py [hunk 11]
```diff
@@ -12,29 +12,23 @@
 
 import os
 from pathlib import Path
+from .settings import *
+import sys
+import mimetypes
 
-# Build paths inside the project like this: BASE_DIR / 'subdir'.
-BASE_DIR = Path(__file__).resolve().parent.parent
+mimetypes.add_type("text/css", ".css", True)
```
Comment (ldeluigi) [inline backend/backend/production_settings.py:19]: Why id you need to write this line? What's the issue with the default?

File: backend/backend/production_settings.py [hunk 12]
```diff
@@ -12,29 +12,23 @@
...
 
+TESTING = sys.argv[1:2] == ['test']
```
Comment (ldeluigi) [inline backend/backend/production_settings.py:21]: I'd suggest using an argument parser here, or better using an environment variable, In any case, why do you need "TESTING" inside the production settings?

File: backend/backend/production_settings.py [hunk 13]
```diff
@@ -12,29 +12,23 @@
...
-ASYNC_GENERATION = True
-
 # Security settings
-ALLOWED_HOSTS = [
-    'commanderspellbook.com',
-    'localhost',
-    '127.0.0.1',
-    'backend.commanderspellbook.com',
-]
+ALLOWED_HOSTS = ['*']
```
Comment (ldeluigi) [inline backend/backend/production_settings.py:31]: This is not a very safe idea because of the presence of classic HTML forms in the Django admin dashboard.
If we want to support custom hosting we could take the allowed host from environment variables? Or else we could simply hardcode it.

This article I found with google explains the need: https://www.divio.com/blog/django-allowed-hosts-explained/

File: backend/backend/production_settings.py [hunk 14]
```diff
@@ -173,6 +71,12 @@
     }
 }
 
+if TESTING:
```
Comment (ldeluigi) [inline backend/backend/production_settings.py:74]: why throttling only under testing?
Anyway, I think throttling should be removed because now kubernetes will serve requests with the load balancer, making randomic the occurrence of throttle errors.

File: backend/docker-compose.yml [hunk 15]
```diff
@@ -6,35 +6,38 @@ services:
       context: .
       target: demo
       args:
-        SECRET_KEY: ${SECRET_KEY:?}
+        SECRET_KEY: tmp_key
```
Comment (ldeluigi) [inline backend/docker-compose.yml:9]: Replace with ${SECRET_KEY:-tmp_key}

Which uses tmp_key as default if SECRET_KEY is not set

File: backend/docker-compose.yml [hunk 16]
```diff
@@ -6,35 +6,38 @@ services:
...
     image: spellbook-backend
-    expose:
-      - 8000
+    ports:
+      - 8000:80
```
Comment (ldeluigi) [inline backend/docker-compose.yml:12]: I think that the container is still using port 8000

File: backend/docker-compose.yml [hunk 17]
```diff
@@ -6,35 +6,38 @@ services:
...
-      SQL_USER: ${DB_USER:?}
-      SQL_PASSWORD: ${DB_PASSWORD:?}
+      SQL_DATABASE: spellbook_db_test
+      SQL_USER: test_user
+      SQL_PASSWORD: test_password
       SQL_HOST: db
       SQL_PORT: 5432
       DATABASE: postgres
     restart: always
-  nginx:
```
Comment (ldeluigi) [inline backend/docker-compose.yml:26]: Why the need of altering nginx here?

File: backend/docker-compose.yml [hunk 18]
```diff
@@ -43,9 +46,14 @@ services:
...
     environment:
       PGPORT: 5432
-      POSTGRES_USER: ${DB_USER:?}
-      POSTGRES_PASSWORD: ${DB_PASSWORD:?}
-      POSTGRES_DB: spellbook_db_prod
+      POSTGRES_USER: test_user
+      POSTGRES_PASSWORD: test_password
+      POSTGRES_DB: spellbook_db_test
+    healthcheck:
+      test: ["CMD-SHELL", "pg_isready", "-q", "-d", "spellbook_db_test", "-U", "test_user" ]
```
Comment (ldeluigi) [inline backend/docker-compose.yml:53]: very cool

File: backend/nginx.conf
Comment (ldeluigi) [inline backend/nginx.conf:1]: Why was this file moved in the parent folder?

**Review by ldeluigi — CHANGES_REQUESTED** [review]
I reviewed every single line of the PR, thanks for all your work!

File: backend/.dockerignore (hunk 1 above)
Comment (accassid) [reply backend/.dockerignore:10]: Yeah, so you're right, they're not necessary for the final image. However, if you'll note for the test step of the CI, rather than running it on the CI machine directly, I'm running it through the docker container with `exec` so that it matches the exact production environment (OS, installed programs, etc). So the tests need to at least be in the container during the build step. We could, in theory, exclude the test files from the image and then in the test step of the CI add them in with docker. But given the negligible size of the test files, I would not prioritize that super high.

File: backend/.kubernetes/kubernetes.yaml (hunk 3 above)
Comment (accassid) [reply backend/.kubernetes/kubernetes.yaml:26]: It's a private repository so hiding the link would only add obscurity, the real security is that it requires auth. I could certainly be convinced to consider making a public image, but it would require a decent bit of work to extract the configuration out in a way that would be useful to anyone else. Right now the image kind of relies on the infrastructure looking exactly like ours.

File: backend/.kubernetes/kubernetes.yaml (hunk 4 above)
Comment (accassid) [reply backend/.kubernetes/kubernetes.yaml:117]: There is now an nginx running on the api image itself as well, with the same configuration for static files as you had before. So for now that will serve the static files. We could entertain the notion of instead pushing the files to their own CDN as part of our build process and have the CDN redirect cdn routes there. However, I think long term we wanted to offload the form work from the django admin panel to a form on the client, so I'm happy with it the way it is for now.

File: backend/docker-compose.yml (hunk 16 above)
Comment (accassid) [reply backend/docker-compose.yml:12]: There is nginx running directly on the pod for ease of use with kubernetes. So the nginx on the container is indeed port 80 (it passes along the gunicon at 8000).

File: backend/docker-compose.yml (hunk 17 above)
Comment (accassid) [reply backend/docker-compose.yml:26]: Putting it directly on the image for the api makes our lives a lot easier with kubernetes since it means that every pod will only need to run one image.

File: backend/nginx.conf
Comment (accassid) [reply backend/nginx.conf:1]: It's now the only nginx configuration needed and it's now directly in the API image, so having it alone in a directory felt excessive.

File: backend/Dockerfile (hunk 6 above)
Comment (accassid) [reply backend/Dockerfile:54]: Great question. I was changing a lot trying to get things to work, so I'll have to double check if this was necessary.

File: backend/Dockerfile (hunk 8 above)
Comment (accassid) [reply backend/Dockerfile:64]: Nginx is in this pod now so that's the port that the outside is going to use.

File: backend/Dockerfile (hunk 7 above)
Comment (accassid) [reply backend/Dockerfile:60]: First off, I'm not sure why I removed this. Maybe because I saw it being run in the demo command. I'll double check what is actually being run. 

As far as offloading serving the static files, I'll echo what I said in another comment, that if we're looking to offload our form work to the frontend, it's not worth our time doing small optimizations for django static files. The average user will never touch a django static file.

File: backend/Dockerfile (hunk 9 above)
Comment (accassid) [reply backend/Dockerfile:66]: From the docs for that property:
```
A positive integer generally in the 2-4 x $(NUM_CORES) range. You’ll want to vary this a bit to find the best for your particular application’s work load.
```
On a pod the number of cores is 1, so we're in the 2-4 range, so 3 is the average. (The default is 1, so we do actually want to overwrite it). We could certainly make it a variable, but it shouldn't actually ever change.

File: backend/Dockerfile (hunk 10 above)
Comment (accassid) [reply backend/Dockerfile:75]: For convenience, they are now sharing an image.

File: backend/Dockerfile.kubernetes
Comment (accassid) [reply backend/Dockerfile.kubernetes:1]: It is not needed at all. Cleaned up.

File: backend/backend/production_settings.py (hunk 11 above)
Comment (accassid) [reply backend/backend/production_settings.py:19]: For whatever reason, with the new setup, the CSS static file were not receiving the correct mime types and django wasn't applying them even though it was fetching them. I'm not actually sure why.

File: backend/backend/production_settings.py (hunk 12 above)
Comment (accassid) [reply backend/backend/production_settings.py:21]: First, this is specifically here to flag whenever the command `python manage.py test` is being run, so the check is accurate. We certainly _could_ use an environment variable, but that would add boilerplate to every time we run the test command.

As for why it's necessary in the production settings, those settings are used when making the production image. On the CI, we actually test directly against that image for the most accurate results.

File: backend/backend/production_settings.py (hunk 13 above)
Comment (accassid) [reply backend/backend/production_settings.py:31]: True, it is safer. I was getting some weird errors with it before, let me see if I can find a better solution.

File: backend/backend/production_settings.py (hunk 14 above)
Comment (accassid) [reply backend/backend/production_settings.py:74]: I wouldn't remove it entirely, it's still some nice last line of defense if we're really getting hammered by a script.

Also, this isn't only setting it for testing, the lines above it still set it for the normal environment. This if statement _overwrites_ what's there only for the testing environment (so the tests don't throttle themselves).

File: backend/.dockerignore (hunk 1 above)
Comment (ldeluigi) [reply backend/.dockerignore:10]: Why aren't you trusting the portability of python, and we need to test it inside the production container? With compiled languages, it's normal to compile and run tests and then package the release version separately and ship it afterwards (I'm thinking about java and C#).

Comment (ldeluigi) [reply backend/.dockerignore:10]: Here inside GitHub Actions, we can test the code on windows, ubuntu and macos. Then, we can fire a docker cp command and get the tests inside the container too if you feel it necessary, The reason of my interest is that I like the idea of having a state-of-the-art docker image, but I don't mind putting the question apart for now.

File: .github/workflows/ci.yml
Comment (ldeluigi) [reply .github/workflows/ci.yml:1]: I'll take care of the refactor in my next free time. I'll open a PR to your branch, so that we cascade from merging that into merging this one eventually.

Comment (ldeluigi) [reply .github/workflows/ci.yml:1]: I'll also add explanation of everything inside some king of "contributors" living documentation I'll begin to write.

File: backend/.kubernetes/kubernetes.yaml (hunk 3 above)
Comment (ldeluigi) [reply backend/.kubernetes/kubernetes.yaml:26]: Yeah, we'll think about this in the future.

File: backend/.kubernetes/kubernetes.yaml (hunk 4 above)
Comment (ldeluigi) [reply backend/.kubernetes/kubernetes.yaml:117]: I can understand your poiny, but I must say that through my experience with Django admin it would be very, very expensive to trying to replicate all those features I added on the client. Surely we will have forms, one day, for users to subit their combo suggestions from the client. But I admit I'm doubting we will ever replace the admin dashboard.

If you agree, I'd think of a simpler, cleaner solution to the problem, and I'm quoting my last discord message:
>  For instance, I think that django itself could serve those or we could even have a single, separate nginx instance that does that maybe

File: backend/docker-compose.yml (hunk 16 above)
Comment (ldeluigi) [reply backend/docker-compose.yml:12]: Thanks for your explanations

File: .github/workflows/ci.yml
Comment (ldeluigi) [reply .github/workflows/ci.yml:1]: Done

**Comment by ldeluigi** [comment]
# Checklist
- [ ] Set up cron jobs
  - [ ] [OPTIONAL, editors can simply stop using google sheets] import_combos
  - [ ] [OPTIONAL, editors have a button to fire this job] export_variants
- [ ] Set up "on release" jobs
  - [ ] seed_website_properties
  - [ ] clean_jobs
- [ ] Merge the sem-ver PR

**Review by ldeluigi — APPROVED** [review]

File: backend/Dockerfile (hunk 6 above)
Comment (ldeluigi) [reply backend/Dockerfile:54]: ok

**Review by ldeluigi — APPROVED** [review]

**Comment by ldeluigi** [comment]
I'm leaving here 3 points for further investigation:
1. CSS mime types in production settings (still needed?)
2. running the application with USER "app" instead of root inside the container
3. collecting static files by default on docker build (was commented)
